fhevm/fhelib.go

```diff
@@ -0,0 +1,318 @@
+var signatureFheRand = makeKeccakSignature("fheRand(bytes1)")
+var signatureFheRandBounded = makeKeccakSignature("fheRandBounded(uint256,bytes1)")
+var signatureCast = makeKeccakSignature("cast(uint256,bytes1)")
+var signatureFheLe = makeKeccakSignature("fheLe(uint256,uint256,bytes1)")
+var signatureFheLt = makeKeccakSignature("fheLt(uint256,uint256,bytes1)")
+var signatureFheEq = makeKeccakSignature("fheEq(uint256,uint256,bytes1)")
+var signatureFheGe = makeKeccakSignature("fheGe(uint256,uint256,bytes1)")
+var signatureFheGt = makeKeccakSignature("fheGt(uint256,uint256,bytes1)")
+var signatureFheShl = makeKeccakSignature("fheShl(uint256,uint256,bytes1)")
+var signatureFheShr = makeKeccakSignature("fheShr(uint256,uint256,bytes1)")
+var signatureFheNe = makeKeccakSignature("fheNe(uint256,uint256,bytes1)")
+var signatureFheNeg = makeKeccakSignature("fheNeg(uint256)")
+var signatureFheNot = makeKeccakSignature("fheNot(uint256)")
+var signatureFheBitAnd = makeKeccakSignature("fheBitAnd(uint256,uint256,bytes1)")
+var signatureFheBitOr = makeKeccakSignature("fheBitOr(uint256,uint256,bytes1)")
+var signatureFheBitXor = makeKeccakSignature("fheBitXor(uint256,uint256,bytes1)")
+var signatureFheIfThenElse = makeKeccakSignature("fheIfThenElse(uint256,uint256,uint256)")
+var signatureFhePubKey = makeKeccakSignature("fhePubKey(bytes1)")
+var signatureTrivialEncrypt = makeKeccakSignature("trivialEncrypt(uint256,bytes1)")
+var signatureDecrypt = makeKeccakSignature("decrypt(uint256)")
+var signatureReencrypt = makeKeccakSignature("reencrypt(uint256,uint256)")
+var signatureVerifyCiphertext = makeKeccakSignature("verifyCiphertext(bytes)")
+var signatureOptimisticRequire = makeKeccakSignature("optimisticRequire(uint256)")
+
+type fheLibPrecompileMethod struct {
+	name                string
+	requiredGasFunction func(environment EVMEnvironment, input []byte) uint64
+	runFunction         func(environment EVMEnvironment, caller common.Address, addr common.Address, input []byte, readOnly bool) ([]byte, error)
+}
+
+func (fheLibMethod *fheLibPrecompileMethod) Name() string {
+	return fheLibMethod.name
+}
+
+func (fheLibMethod *fheLibPrecompileMethod) RequiredGas(environment EVMEnvironment, input []byte) uint64 {
+	return fheLibMethod.requiredGasFunction(environment, input)
+}
+
+func (fheLibMethod *fheLibPrecompileMethod) Run(environment EVMEnvironment, caller common.Address, addr common.Address, input []byte, readOnly bool) ([]byte, error) {
+	return fheLibMethod.runFunction(environment, caller, addr, input, readOnly)
+}
+
+// Mapping between function signatures and the functions to call
+var signatureToFheLibMethod = map[uint32]*fheLibPrecompileMethod{}
+
+func GetFheLibMethod(signature uint32) (fheLibMethod *fheLibPrecompileMethod, found bool) {
+	fheLibMethod, found = signatureToFheLibMethod[signature]
+	return
+}
+
+func init() {
+	signatureToFheLibMethod[signatureFheAdd] = &fheLibPrecompileMethod{
```

Comment:
Maybe we could also get rid of signatureFheAdd variables? It could be little more error prone and more things to remember to change when copy pasting (map key could be different from value)

Say there could be another field in fheLibPrecompileMethod, `arguments`, and it could contain `(uint256,uint256,bytes1)` and signature could be derived by method contatenating `name+arguments)`.

Then we define a global list of such methods like:
```
var fheLibMethods = []*fheLibPrecompileMethod{
    ...
    &fheLibPrecompileMethod{
		name:                "fheSub",
                arguments:           "(uint256,uint256,bytes1)",
		requiredGasFunction: fheAddSubRequiredGas,
		runFunction:         fheSubRun,
    },
    ...
}
```

And also maybe this entire map may be returned by a separate function which would be called inside init.

Then inside the function we could have a short loop that goes through and indexes the map
```
func generatePrecompileMap() map[uint32]*fheLibPrecompileMethod{} {
  var result map[uint32]*fheLibPrecompileMethod{}
  
  for _, method := range(fheLibMethods) {
    // possibly add validations if duplicate signatures are defined to panic early
    // and other checks for data
    result[method.ethSignatureBytes()] = method;
  }
  
  return result
}
```

Reply:
I like it! I will push a change with this then merge